overlord/ifacestate/ifacestate_test.go

```diff
@@ -8086,21 +8110,190 @@ func (s *interfaceManagerSuite) TestPreseedAutoConnectErrorWithInterfaceHooks(c
+}
+
+func (s *interfaceManagerSuite) TestPreseedAutoConnectInternalErrorOnMarkPreseededState(c *C) {
+	restore := snapdenv.MockPreseeding(true)
+	defer restore()
+
+	autoConnectTask, markPreseededTask := s.autoconnectChangeForPreseeding(c, false)
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	markPreseededTask.SetStatus(state.DoingStatus)
+	st.Unlock()
 	s.settle(c)
+	s.state.Lock()
+
+	c.Check(strings.Join(autoConnectTask.Log(), ""), Matches, `.* internal error: unexpected state of mark-preseeded task: Doing`)
+}
 
+func (s *interfaceManagerSuite) TestPreseedAutoConnectInternalErrorMarkPreseededMissing(c *C) {
+	restore := snapdenv.MockPreseeding(true)
+	defer restore()
+
+	skipMarkPreseeded := true
+	autoConnectTask, markPreseededTask := s.autoconnectChangeForPreseeding(c, skipMarkPreseeded)
+	c.Assert(markPreseededTask, IsNil)
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	st.Unlock()
+	s.settle(c)
 	s.state.Lock()
-	defer s.state.Unlock()
 
-	// Ensure that the task succeeded.
-	c.Check(change.Status(), Equals, state.ErrorStatus)
-	c.Check(change.Err(), ErrorMatches, `cannot perform the following tasks:\n.*interface hooks are not yet supported in preseed mode.*`)
+	c.Check(strings.Join(autoConnectTask.Log(), ""), Matches, `.* internal error: mark-preseeded task not found in preseeding mode`)
+}
+
+func (s *interfaceManagerSuite) TestFirstTaskAfterBootWhenPreseedingErrors(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "")
+
+	setupTask := st.NewTask("some-task", "")
+	setupTask.Set("snap-setup", &snapstate.SnapSetup{SideInfo: &snap.SideInfo{RealName: "test-snap"}})
+	chg.AddTask(setupTask)
+
+	markPreseeded := st.NewTask("fake-mark-preseeded", "")
+	markPreseeded.WaitFor(setupTask)
+	_, err := ifacestate.FirstTaskAfterBootWhenPreseeding("test-snap", markPreseeded)
+	c.Check(err, ErrorMatches, `internal error: fake-mark-preseeded task not in change`)
+
+	chg.AddTask(markPreseeded)
+
+	_, err = ifacestate.FirstTaskAfterBootWhenPreseeding("test-snap", markPreseeded)
+	c.Check(err, ErrorMatches, `internal error: cannot find start-setup-profiles or install hook for snap "test-snap"`)
+
+	task1 := st.NewTask("some-task", "")
+	task1.WaitFor(markPreseeded)
+	chg.AddTask(task1)
+
+	_, err = ifacestate.FirstTaskAfterBootWhenPreseeding("test-snap", markPreseeded)
+	c.Check(err, ErrorMatches, `internal error: cannot find start-setup-profiles or install hook for snap "test-snap"`)
+
+	task2 := st.NewTask("run-hook", "")
+	hsup := hookstate.HookSetup{Hook: "install", Snap: "other-snap"}
+	task2.Set("hook-setup", &hsup)
+	task2.WaitFor(task1)
+	chg.AddTask(task2)
+	_, err = ifacestate.FirstTaskAfterBootWhenPreseeding("test-snap", markPreseeded)
+	c.Check(err, ErrorMatches, `internal error: cannot find start-setup-profiles or install hook for snap "test-snap"`)
+
+	task3 := st.NewTask("start-snap-services", "")
+	task3.Set("snap-setup-task", setupTask.ID())
+	task2.WaitFor(task2)
```

Comment:
task2, task2 this looks wrong but the code also isn't checking for wait relationshipts, just names

Reply:
Fixed, although wait relationships are irrelevant for this test as it only verifies FirstTaskAfterBootWhenPreseeding helper (it would become relevant if HaltTasks was used per your other comment).

Wait relationships are tested in `TestPreseedAutoConnectWithInterfaceHooks` test.